### ISSUE TITLE
Issue identified : Something really wrong happened!

### ISSUE THREAD

**Issue by realogic** [gillesvs/librelink#39]
** Which country are you from **
ITALY

**Describe the bug**
Since this morning, the add-on fails with error "Something really wrong happened!"

**Log file**
Logger: homeassistant.config_entries
Source: config_entries.py:640
First occurred: 11:55:48 (2 occurrences)
Last logged: 11:55:57

Error setting up entry ***MAIL*** for librelink
Traceback (most recent call last):
  File "/config/custom_components/librelink/api.py", line 200, in api_wrapper
    response.raise_for_status()
    ~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/usr/local/lib/python3.13/site-packages/aiohttp/client_reqrep.py", line 1157, in raise_for_status
    raise ClientResponseError(
    ...<5 lines>...
    )
aiohttp.client_exceptions.ClientResponseError: 430, message='', url='https://api-eu.libreview.io/llu/auth/login'

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 640, in __async_setup_with_context
    result = await component.async_setup_entry(hass, self)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/librelink/__init__.py", line 48, in async_setup_entry
    sessionToken = await myLibrelinkLogin.async_get_token()
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/librelink/api.py", line 148, in async_get_token
    reponseLogin = await api_wrapper(
                   ^^^^^^^^^^^^^^^^^^
    ...<9 lines>...
    )
    ^
  File "/config/custom_components/librelink/api.py", line 209, in api_wrapper
    raise LibreLinkApiCommunicationError(
        "Error fetching information",
    ) from exception
custom_components.librelink.api.LibreLinkApiCommunicationError: Error fetching information

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Additional context**
Add any other context about the problem here.

**Comment by D0ublek1ll** [comment]
I started having the same issue an hour ago, something seems to have changed on the libreview side of things as their API is returning an HTTP/430 status, which is not officially defined.

For me, this started happening at 11:04 today (CET).

**Comment by realogic** [comment]
Yes, very same time

**Comment by TekFan** [comment]
I can confirm this happened to me as well at 11:04 CET today.
Their server could simply be down.

**Comment by D0ublek1ll** [comment]
I found another issue regarding the same status code from a couple of months ago: https://github.com/gillesvs/librelink/issues/31

In that it was suggested that it could be related to having to accept the terms of service on the libreview site/librelinkup app.
I am trying that right now, will update with results.

Edit: Did not fix the issue.

**Comment by realogic** [comment]
I tried already, signing in and out the LibreLinkUp on my mobile. Accepted the terms, reloaded the add-on, no changes
On app data is retrieved, so I don't think the service is down

**Comment by lucamargoni** [comment]
Same problem. In these days the app LibrelinkUp warned to update compulsorily, otherwise it would stop working. They probably changed something in the API or simply the server for italy area

**Comment by D0ublek1ll** [comment]
Libreview, librelinkup and freestyle libre all seem to work as intended, so it seems to be the Libreview API that is causing these issues.

I tried using different regions/api servers but all give the same result.
Maybe @gillesvs can give us more insight into what could cause this issue?

**Comment by Pleister023** [comment]
Same here, same timestamp:


This error originated from a custom integration.

Logger: custom_components.librelink
Source: helpers/update_coordinator.py:414
integration: LibreLink integration (documentation, issues)
First occurred: 11:03:56 AM (1 occurrences)
Last logged: 11:03:56 AM

Error fetching librelink data: Something really wrong happened!

**Comment by gui-dos** [comment]
They are now requiring a minimal version `4.12.0` to be specified in https://github.com/gillesvs/librelink/blob/0b12fa9c415c7c49d0c235f6fa881501eb6409b3/custom_components/librelink/const.py#L41 and a new "Account-Id" HTTP header carrying the SHA256 digest of a user's `id` as a 64-char hexadecimal string along the JWT token.

Look at my Swift one-liners I PRed with https://github.com/JohanDegraeve/xdripswift/commit/13e4a1e7418364fbdb68e5bd54cac888fd2eb9b8 ;-)

**Comment by Pleister023** [comment]
_> They are now requiring a minimal version `4.12.0` to be specified in
> 
> https://github.com/gillesvs/librelink/blob/0b12fa9c415c7c49d0c235f6fa881501eb6409b3/custom_components/librelink/const.py#L41
> 
> and a new "Account-Id" HTTP header carrying the SHA256 digest of `user-id` as a 64-char hexadecimal string along the JWT token.
> Look at my Swift one-liners I PRed with [JohanDegraeve/xdripswift@13e4a1e](https://github.com/JohanDegraeve/xdripswift/commit/13e4a1e7418364fbdb68e5bd54cac888fd2eb9b8) ;-)_

Thanks for the update. I managed to change 4.7 into 4.12.0 in that config file. 
I have no clue what to do with the rest of your solution: "and a new "Account-Id" HTTP header carrying the SHA256 digest of user-id as a 64-char hexadecimal string along the JWT token."
The link below that brings me to a lot of code. No idea what to do with this. Any help is much appreciated.

Thanks.

**Comment by Zpuit** [comment]
Have the issue as well… latest reported glucose measurement sensor value at at 11:03 CET

**Comment by xamrex** [comment]
Same issue here @gillesvs are you planing to fix issue ?

**Comment by padoubek** [comment]
Also the same issue, Czech region. However the region seems to be not important in this case. Please, prepare uodate to fix it.

**Comment by Flemming789** [comment]
Same issue here in Denmark.

**Comment by santypav** [comment]
Spain is not different

**Comment by WillemBastiaens** [comment]
Belgium has the same problem since 11:04 CET

**Comment by xamrex** [comment]
based on https://github.com/gillesvs/librelink/issues/39#issuecomment-2531500325
i noticed that in const.py
`VERSION_APP = "4.7"`
should be changed to
`VERSION_APP = "4.11.0"`

but unfortunately i have no idea what to do with this sha256 function ;(

**Comment by xamrex** [comment]
@gui-dos 
can you please take a look at 
[                "Application": APPLICATION,](https://github.com/gillesvs/librelink/blob/0b12fa9c415c7c49d0c235f6fa881501eb6409b3/custom_components/librelink/api.py#L50)
as i understand I need to add bellow: this line

"Account-Id": **SOMETHING** // and that SOMETHING has to be in sha256()
but what should be set as Account-id? 
is it just sha256("4.11.0") ?

**Comment by gui-dos** [comment]
@xamrex:

The `Account-Id` header has to be appended to the headers defined in https://github.com/gillesvs/librelink/blob/0b12fa9c415c7c49d0c235f6fa881501eb6409b3/custom_components/librelink/api.py#L99-L110

If you follow yourself you can reuse the `patient_id`, otherwise you have to extract the logging user's `id` from `reponseLogin["data"]["user"]["id"]` (https://github.com/gillesvs/librelink/blob/0b12fa9c415c7c49d0c235f6fa881501eb6409b3/custom_components/librelink/api.py#L168).

I just noticed tonight's Python implementation by https://github.com/robberwick/pylibrelinkup/commit/06bce4537a84557e2a05c10e2409c963bdf17291 which makes use of 
`hashlib.sha256(account_id.encode()).hexdigest()` ;-)

`4.12.0` version must be used instead of `4.11.0` now, of course.

**Comment by xamrex** [comment]
@gui-dos 
I got this code:

`class LibreLinkApiClient:


    Attributes:
        token: The long life token to authenticate.
        base_url: For API calls depending on your location.
        session: aiohttp.ClientSession object for the open session.
        hashed_libreLinkID: The hashed LibreLinkID.
    """

    def __init__(
        self, token: str, base_url: str, session: aiohttp.ClientSession, hashed_libreLinkID: str
    ) -> None:
        """Sample API Client."""
        self._token = token
        self._session = session
        self._hashed_libreLinkID = hashed_libreLinkID  # Store it as an instance variable
        self.connection_url = base_url + CONNECTION_URL

    async def async_get_data(self) -> any:
        """Get data from the API."""
        APIreponse = await api_wrapper(
            self._session,
            method="get",
            url=self.connection_url,
            headers={
                "product": PRODUCT,
                "version": VERSION_APP,
                "Application": APPLICATION,
                "Authorization": "Bearer " + self._token,
                "Account-id": self._hashed_libreLinkID,  # Access the instance variable here
            },
            data={},
        )

        # Log the API status
        _LOGGER.debug(
            "Return API Status:%s ",
            APIreponse["status"],
        )

        # API status returns 0 if everything goes well.
        if APIreponse["status"] == 0:
            patients = sorted(APIreponse["data"], key=lambda x: x["patientId"])
        else:
            patients = APIreponse  # For debugging in case the status is not OK

        _LOGGER.debug(
            "Number of patients: %s and patient list: %s",
            len(patients),
            patients,
        )

        return patients`

But unfortunately it returns:
:Return API Status:{'status': 0, 'data': [], 'ticket': {'token': 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6ImFlYjVhMGMxLWFiY2QtMTFlZC05MDhlLTAyNDJhYzExMDAwYSIsImZpcnN0TmFtZSI6IkFydHVyIiwibGFzdE5hbWUiOiJTemFyYW5pZWMiLCJjb3VudHJ5IjoiUEwiLCJyZWdpb24iOiJldSIsInJvbGUiOiJwYXRpZW50IiwidW5pdHMiOjEsInByYWN0aWNlcyI6W10sImMiOjEsInMiOiJsbHUuYW5kcm9pZCIsImV4cCI6MTc0OTQyMzI4N30.h600H_q5_kkgpiTcekQbYNZRoOBQPDDZRCHANGED', 'expires': 1749423287, 'duration': 15552000000}} 

And if I m right I sould have PatientID insinde data[], right?
any idea why its empty?

**Comment by xamrex** [comment]
To ALL
**Here is working fork:**

https://github.com/kubasaw/librelink

**Comment by syska** [comment]
Also got the sam issue.

Happy to see there might be a solution soon to this issues.

@xamrex 

First i was like ... there is alot of changes. Then I looked closer ... 

You changed alot, but I just read the comment from the author on you commit.

Is is possible you can make a PR to just fix this issue so the packages can be updated via HACS?

**Comment by xamrex** [comment]
@syska its not my fork ;-)

> Is is possible you can make a PR to just fix this issue so the packages can be updated via HACS?

Im afriad no,
It looks like author of this repo has some issue with PR:
Here is one opened 
https://github.com/gillesvs/librelink/pull/34
and not merged.


So the easiest way to fix issue is to remove this repo form HACS and add this one:
https://github.com/kubasaw/librelink

**Comment by D0ublek1ll** [comment]
It appears this new repo (https://github.com/kubasaw/librelink) was added to HACS a couple of hours ago, its using the same name and icon. There are some changes as to which entities it presents and the naming has changed compared to the old integration (<name>_glucose_measurement -> <name>_measurement) 

Also the newly added integration will install in the same location as the old one. so you should first delete and uninstall the old integration before installing the new one.

Given the author of this repo's lack of response and the past issues with the PR on this repo it should probably be removed from hacs.

**Comment by realogic** [comment]
Moved to that repo too, just finished reconfiguring sensors. Thank you!
Still, thanks to the original author @gillesvs to his precious work

**Comment by syska** [comment]
A HUGE thanks to @gillesvs ... 

It's not until things breaks you know how much you really need it.

Now we just need to figure out where it should be hosted going forward.

**Comment by r-jean-pierre** [comment]
Also have "Something really wrong happened!", using from France, the librelinkup on mobile works

**Comment by xamrex** [comment]
@r-jean-pierre  use this custom repo instead:
https://github.com/kubasaw/librelink

**Comment by zaffraz** [comment]
Hello. After the change of repo and restart ha I am now able to log in but then I get this errore 
![Screenshot_20241212-003150_Home Assistant](https://github.com/user-attachments/assets/04441b10-2e15-4476-9fcf-28073e9890a5)

**Comment by syska** [comment]
@zaffraz 

I Can't read that language ... I have abosolutely no idea what is says.

But I would open a terminal or editor and delete all the old "libre" files and try to install again. You should be able to select a "Patinet Id" with a radio button. ( Probably the email you want to follow )

**Comment by TekFan** [comment]
> I Can't read that language ... I have abosolutely no idea what is says.

Luckily, Google Translate can ;-)
"Not all mandatory fields are filled in."

**Comment by xamrex** [comment]
@zaffraz for me it looks like there is no patiend_id,
and its probably beacuse you put **libreview** login and password, instead of **librelink**

Remember to download librelink on phone once, and accept terms.

**Comment by zaffraz** [comment]
You were right. Now works. Thanks everyone for the supporto
> @zaffraz for me it looks like there is no patiend_id,
> and its probably beacuse you put **libreview** login and password, instead of **librelink**
> 
> Remember to download librelink on phone once, and accept terms.

**Comment by Rasmus-nylander** [comment]
Hey Guys. I have tried to install the new https://github.com/kubasaw/librelink Custom repositories in HACS. But I get the following error. I admit I am pretty new in HA, so it might be me doing somehintg worng, but everything else I have installed in HACS have worked well so far. 

Any help would be greatly appreciated :-)  
<img width="431" alt="Skærmbillede 2024-12-30 kl  00 52 58" src="https://github.com/user-attachments/assets/7ba2d591-9584-4251-86f5-4ae57a058da1" />

The only "Libre link" I can find in the HACS store, is the "old version" from gillesvs.

**Comment by xamrex** [comment]
@Rasmus-nylander it looks like you have already added this repo to hacs.
go to HACS, top right corner Custom repo, and remove all integrations that starts with librelink

Then add this repo: https://github.com/kubasaw/librelink
Then in search bar, look for librelink, click that, and on botton there will be Download button.

**Comment by Rasmus-nylander** [comment]
@xamrex Thanks. I think I got confused that both versions seems to be by gillesvs when listed in HACS. But I took notice of the date and stars ratings, and spotted the difference. It all works again. Thanks a lot guys!

**Comment by Finest** [comment]
i addet and download the fork version from https://github.com/kubasaw/librelink and when i try to add the integration ill get

Unknown error occurred

![image](https://github.com/user-attachments/assets/03a1e6af-57b4-43c8-a8cb-559c4eb921ee)
